### PR TITLE
Various small fixes

### PR DESCRIPTION
**Pull request by aseyboldt**
A collection of small fixes I have accumulated...

I thought it would be easier to just submit this as one PR instead of 5 small ones. The individual commits contain one fix each.

### CODE REVIEW

**Comment by ricardoV94** [comment]
Looks neat. Question is https://github.com/pymc-devs/pytensor/pull/375/commits/d5574c1a0138bc2484af01dd05b94c65bc426691 fixing a bug or just providing better static shape?

**Comment by aseyboldt** [comment]
Usually I just got a None shape, but in the code it seemed to just return shape==1 sometimes: https://github.com/pymc-devs/pytensor/commit/d5574c1a0138bc2484af01dd05b94c65bc426691#diff-f5839e5efccae3b80e368b3d43abd60c65d64a012eb92a231c682805d4ab1e62L747

Not sure what that was about...

**Comment by ricardoV94** [comment]
> Not sure what that was about...

We need to know whether the output is broadcastable for the grad. Every thing besides that is just a plus that was not implemented since we added static type shapes I guess.

File: pytensor/tensor/subtensor.py [hunk 1]
```diff
@@ -728,26 +738,39 @@ def make_node(self, x, *inputs):
...
-                            start = 0
-                        if p.stop is None or (
-                            isinstance(p.stop, (int, np.integer, np.ndarray))
-                            and p.stop > start
-                        ):
-                            out_shape.append(1)
-                            continue
-
-                out_shape.append(None)
+                except NotScalarConstantError:
```
Comment (ricardoV94) [inline pytensor/tensor/subtensor.py:755]: Nitpick can we do the 3 in one try/except block?

Comment (aseyboldt) [reply pytensor/tensor/subtensor.py]: done

**Review by ricardoV94 — APPROVED** [review]
Looks good

File: pytensor/tensor/subtensor.py (hunk 1 above)
Comment (aseyboldt) [reply pytensor/tensor/subtensor.py]: Didn't work out after all, we have to accept None values, and that raises an exception.

Comment (ricardoV94) [reply pytensor/tensor/subtensor.py]: Np

**Comment by ricardoV94** [comment]
Tests looked like they were passing before refreshing :)

File: pytensor/tensor/subtensor.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/subtensor.py]: There's also `get_scalar_constant` that may not raise with None. Not sure now

File: pytensor/link/numba/dispatch/basic.py [hunk 2]
```diff
@@ -561,11 +561,18 @@ def {fn_name}({", ".join(input_names)}):
 @numba_funcify.register(Subtensor)
 @numba_funcify.register(AdvancedSubtensor1)
 def numba_funcify_Subtensor(op, node, **kwargs):
-    subtensor_def_src = create_index_func(
-        node, objmode=isinstance(op, AdvancedSubtensor)
-    )
+    objmode = isinstance(op, AdvancedSubtensor)
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/basic.py:564]: Btw, how comes numba doesn't support advanced indexing?

File: pytensor/tensor/shape.py [hunk 3]
```diff
@@ -972,10 +972,7 @@ def make_node(self, x):
         x = as_tensor_variable(x)
         if x.type.ndim <= max(self.axes):
             raise ValueError("Trying to unbroadcast of non-existent dimension")
-        shape = [
-            None if (sh == 1 and i in self.axes) else sh
-            for i, sh in enumerate(x.type.shape)
-        ]
+        shape = [None if i in self.axes else sh for i, sh in enumerate(x.type.shape)]
```
Comment (ricardoV94) [inline pytensor/tensor/shape.py:975]: This is losing more information than before. IIRC it's fine to unbroadcast an axis that's already known to be non-broadcastable.

The perform is actually just a pass through, no checks are done

File: pytensor/tensor/subtensor.py [hunk 4]
```diff
@@ -722,32 +732,51 @@ def make_node(self, x, *inputs):
                     f"Incompatible types for Subtensor template. Expected {input.type}, got {expected_type}."
                 )
 
-        # infer the broadcasting pattern
-        padded = get_constant_idx(
-            self.idx_list, (None,) + inputs, allow_partial=True
-        ) + [slice(None, None, None)] * (x.type.ndim - len(idx_list))
+        padded = [
```
Comment (ricardoV94) [inline pytensor/tensor/subtensor.py:735]: nitpick you may want to call this "enhancement" in the commit prefix. None was a correct static shape output, just not the best we could have done.

File: pytensor/link/numba/dispatch/basic.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/link/numba/dispatch/basic.py:564]: I did some research. It's being improved on

**Comment by fonnesbeck** [comment]
Can this be merged? There's a fix in here for a nutpie bug I keep running into.

**Comment by ricardoV94** [comment]
> Can this be merged? There's a fix in here for a nutpie bug I keep running into.

There's a change that has to be reverted, I'll try to get on it Monday, thanks for pinging

**Review by ricardoV94 — APPROVED** [review]

**Pull request merged**